controllers/storageautoscaler/storageautoscaler_controller.go

```diff
@@ -0,0 +1,255 @@
+	"github.com/prometheus/common/model"
+	ocsv1 "github.com/red-hat-storage/ocs-operator/api/v4/v1"
+	"github.com/red-hat-storage/ocs-operator/v4/controllers/platform"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// StorageAutoscalerReconciler is the reconciler for StorageAutoscaler objects
+// nolint:revive
+type StorageAutoscalerReconciler struct {
+	client.Client
+	ctx               context.Context
+	Log               logr.Logger
+	Scheme            *runtime.Scheme
+	OperatorNamespace string
+	RequeueTime       time.Duration
+	SyncMap           sync.Map
+	Event             chan event.GenericEvent
+}
+
+// SetupWithManager sets up the reconciler with the manager
+func (r *StorageAutoscalerReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// run a go routine to scrape the metrics from prometheus periodically
+	go r.scrapeMetricsPeriodically()
+
+	storageAutoscalerReconcilerController := ctrl.NewControllerManagedBy(mgr).
+		For(&ocsv1.StorageAutoScaler{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		// watch for generic events to trigger the reconcile
+		WatchesRawSource(source.Channel(r.Event,
+			&handler.EnqueueRequestForObject{},
+		))
+
+	return storageAutoscalerReconcilerController.Complete(r)
+}
+
+// +kubebuilder:rbac:groups="monitoring.coreos.com",resources=*,verbs=get
+
+// Reconcile reconciles the StorageAutoscaler object
+func (r *StorageAutoscalerReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	r.ctx = ctx
+	r.Log.Info("reconciling storage autoscaler", "namespace", req.Namespace, "name", req.Name)
+
+	// print the sync map
+	r.SyncMap.Range(func(key, value interface{}) bool {
+		r.Log.Info("main function sync map", "key", key.(string), "value", value.(float64))
+		return true
+	})
+
+	// todo implement the reconcile logic
+
+	return reconcile.Result{}, nil
+}
+
+func (r *StorageAutoscalerReconciler) scrapeMetricsPeriodically() {
+	for {
+		select {
+		// context of the main reconcile loop is cancelled
+		case <-context.Background().Done():
+			r.Log.Info("context cancelled, stopping scraping metrics")
+			return
+
+		case <-time.After(r.RequeueTime):
+			r.reconcileMetrics()
+		}
+	}
+}
+
+func (r *StorageAutoscalerReconciler) reconcileMetrics() {
+	r.Log.Info("scraping metrics")
+	metrics, err := r.scrapeMetrics()
+	if err != nil {
+		r.Log.Error(err, "failed to scrape metrics")
+		return
+	}
+	r.Log.Info("scraped metrics", "metrics", metrics)
+
+	r.Log.Info("update sync map")
+	r.updateSyncMap(metrics)
+	r.Log.Info("sync map updated")
```

Comment:
I see lots and lots of log info messages. I'm assuming these are just for debugging and will be reduced in the final draft?

Reply:
We can make them debug, but as there is no good way to track error in go-routine I like to keep them to understand which part is breaking